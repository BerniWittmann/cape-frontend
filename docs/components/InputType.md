# InputType

A component to edit a value and it's type

## Props

<!-- @vuese:InputType:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|value|The Value of the input field|`String`|`false`|''|
|type|The Type of the input field|`String`|`false`|'String'|
|size|Size of the controls|`'medium'` / `'small'` / `'mini'`|`false`|'normal'|
|placeholder|The Placeholder for the input field|`String`|`false`|undefined|
|typePlaceholder|The Placeholder for the type field|`String`|`false`|undefined|
|showTypeSelect|Display the select to change the type|`Boolean`|`false`|true|

<!-- @vuese:InputType:props:end -->


## Events

<!-- @vuese:InputType:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|change|Emitted when the input value or type is changed|data object with value and type|

<!-- @vuese:InputType:events:end -->


## Methods

<!-- @vuese:InputType:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|validate|validates the form|Callback function with the result as boolean|

<!-- @vuese:InputType:methods:end -->


