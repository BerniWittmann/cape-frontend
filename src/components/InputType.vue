<template>
  <el-form :model="data" :rules="rules" ref="inputTypeForm" :size="size">
    <el-form-item prop="value">
      <el-input :placeholder="placeholder" v-model="data.value" class="input-with-select" @change="handleChange" :size="size">
        <el-select v-model="data.type" slot="prepend" :placeholder="typePlaceholder" @change="handleChange" v-if="showTypeSelect" :size="size">
          <el-option v-for="option in typeOptions" :key="option.value" :label="option.label"
                     :value="option.value"></el-option>
        </el-select>
      </el-input>
    </el-form-item>
  </el-form>
</template>

<script>
/*
 * @vuese
 * @group Components
 *
 * A component to edit a value and it's type
 */
export default {
  name: 'InputType',
  model: {
    prop: 'data',
    event: 'change'
  },

  props: {
    // The Value of the input field
    value: {
      type: String,
      default: ''
    },
    // The Type of the input field
    type: {
      type: String,
      default: 'String'
    },
    // Size of the controls
    size: {
      // `'medium'` / `'small'` / `'mini'`
      type: String,
      default: 'normal'
    },
    // The Placeholder for the input field
    placeholder: {
      type: String,
      default: undefined
    },
    // The Placeholder for the type field
    typePlaceholder: {
      type: String,
      default: undefined
    },
    // Display the select to change the type
    showTypeSelect: {
      type: Boolean,
      default: true
    }
  },

  data() {
    return {
      data: {
        value: undefined,
        type: undefined
      },
      typeOptions: [{
        value: 'String',
        label: this.$t('types.string')
      }, {
        value: 'Number',
        label: this.$t('types.number')
      }, {
        value: 'Boolean',
        label: this.$t('types.boolean')
      }],
      rules: {
        value: [{
          validator: this.validateValue, trigger: 'blur'
        }]
      }
    }
  },

  watch: {
    value: {
      deep: true,
      immediate: true,
      handler: function (newValue) {
        this.data.value = newValue
      }
    },
    type: {
      deep: true,
      immediate: true,
      handler: function (newValue) {
        this.data.type = newValue
      }
    }
  },

  methods: {
    handleChange() {
      this.$refs.inputTypeForm.validate((valid) => {
        if (valid) {
          this.$emit('change', this.data)
        }
      })
    },

    // @vuese
    // validates the form
    // @arg Callback function with the result as boolean
    validate(cb) {
      this.$refs.inputTypeForm.validate(cb)
    },

    validateValue(rule, value, callback) {
      if (!value || value.length === 0) return callback(new Error(this.$t('input_type.required')))

      if (this.data.type === 'Number') {
        if (isNaN(value)) return callback(new Error(this.$t('input_type.number')))
      } else if (this.data.type === 'Boolean') {
        if (!['TRUE', 'FALSE'].includes(value)) return callback(new Error(this.$t('input_type.boolean')))
      }

      callback()
    }
  }
}

</script>

<style lang="scss">

.input-with-select {
  .el-select .el-input {
    width: 100px;
  }
}
</style>
