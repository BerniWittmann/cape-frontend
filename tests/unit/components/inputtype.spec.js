import { mount, config } from '@vue/test-utils'
import { i18n, monkeyPatchTransitions } from '../setupPlugins'

import InputType from '@/components/InputType.vue'

describe('Components', () => {
  describe('InputType', () => {
    monkeyPatchTransitions()

    config.transition = false
    let propsData
    let cmp

    beforeEach(() => {
      propsData = {
        value: 'Test Text',
        type: 'String'
      }
      // console.warn = jest.fn()
      render()
    })

    function render(stubs = {}) {
      cmp = mount(InputType, {
        i18n,
        propsData,
        stubs: {
          transition: false,
          ...stubs
        }
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('has a prop to hide the type select', () => {
      propsData.showTypeSelect = false
      render()

      expect(cmp.html()).toMatchSnapshot()
      expect(cmp.contains('.el-select')).toBeFalsy()
    })

    it('has a prop to control the size', () => {
      propsData.size = 'small'
      render()

      expect(cmp.html()).toMatchSnapshot()
      expect(cmp.contains('.el-form-item--small')).toBeTruthy()
      expect(cmp.contains('.el-select--small')).toBeTruthy()
      expect(cmp.contains('.el-input--small')).toBeTruthy()
    })

    describe('it can change the type', () => {
      it('displays a select to change the type', () => {
        const select = cmp.find('.el-select')
        expect(select.exists()).toBeTruthy()
        const options = select.findAll('.el-select-dropdown__item')
        expect(options.at(0).text()).toEqual('types.string')
        expect(options.at(1).text()).toEqual('types.number')
        expect(options.at(2).text()).toEqual('types.boolean')
      })
      describe('it can change the type to a string', () => {
        it('can change the type to string', () => {
          propsData.type = 'Number'
          render()

          expect(cmp.vm.data.type).not.toEqual('String')

          cmp.find('.el-select-dropdown__item').trigger('click')

          expect(cmp.vm.data.type).toEqual('String')
        })

        describe('validates the field on change', () => {
          beforeEach(() => {
            propsData.type = 'String'
            render()
          })
          it('shows no error if no input given', (done) => {
            const input = cmp.findAll('input').at(1)
            expect(input.exists()).toBeTruthy()
            input.setValue('')
            input.trigger('blur')
            input.trigger('change')

            cmp.vm.$nextTick(() => {
              expect(cmp.html()).toMatchSnapshot()
              const err = cmp.find('.el-form-item__error')
              expect(err.exists()).toBeFalsy()
              expect(cmp.emitted().change[0][0]).toEqual({
                type: 'String',
                value: ''
              })
              done()
            })
          })
          it('emits change event if valid', () => {
            const input = cmp.findAll('input').at(1)
            expect(input.exists()).toBeTruthy()
            input.setValue('my New Value')
            input.trigger('change')

            expect(cmp.emitted().change[0][0]).toEqual({
              type: 'String',
              value: 'my New Value'
            })
          })
        })
      })

      describe('it can change the type to a boolean', () => {
        it('can change the type to boolean', () => {
          propsData.type = 'String'
          render()

          expect(cmp.vm.data.type).not.toEqual('Boolean')

          cmp.findAll('.el-select-dropdown__item').at(2).trigger('click')

          expect(cmp.vm.data.type).toEqual('Boolean')
        })

        it('renders a select field', () => {
          propsData.type = 'Boolean'
          render()

          expect(cmp.html()).toMatchSnapshot()
          const selectGroup = cmp.find('.select-group')
          expect(selectGroup.exists()).toBeTruthy()
          expect(selectGroup.findAll('.el-select').length).toEqual(2)
          const select = selectGroup.findAll('.el-select').at(1)
          expect(select.exists()).toBeTruthy()
          const options = select.findAll('.el-select-dropdown__item')
          expect(options.length).toEqual(2)
          expect(options.at(0).props('value')).toEqual('TRUE')
          expect(options.at(0).props('label')).toEqual('TRUE')
          expect(options.at(0).text()).toEqual('TRUE')
          expect(options.at(1).props('value')).toEqual('FALSE')
          expect(options.at(1).props('label')).toEqual('FALSE')
          expect(options.at(1).text()).toEqual('FALSE')
        })

        describe('validates the field on change', () => {
          beforeEach(() => {
            propsData.type = 'Boolean'
            render()
          })
          it('shows no error if no input given', (done) => {
            const input = cmp.findAll('.el-select').at(1)
            expect(input.exists()).toBeTruthy()
            cmp.vm.data.value = ''
            input.vm.$emit('change')

            cmp.vm.$nextTick(() => {
              expect(cmp.html()).toMatchSnapshot()
              const err = cmp.find('.el-form-item__error')
              expect(err.exists()).toBeFalsy()
              expect(cmp.emitted().change).toBeTruthy()
              expect(cmp.emitted().change[0][0]).toEqual({
                type: 'Boolean',
                value: ''
              })
              done()
            })
          })
          it('shows error if invalid input', (done) => {
            render()
            const input = cmp.findAll('.el-select').at(1)
            expect(input.exists()).toBeTruthy()
            cmp.vm.data.value = 'F4LsE'
            input.vm.$emit('change')

            cmp.vm.$nextTick(() => {
              expect(cmp.html()).toMatchSnapshot()
              const err = cmp.find('.el-form-item__error')
              expect(err.exists()).toBeTruthy()
              expect(err.text()).toEqual('input_type.boolean')
              expect(cmp.emitted().change).toEqual(undefined)
              done()
            })
          })
          it('emits change event if value is FALSE', () => {
            const input = cmp.findAll('.el-select').at(1)
            expect(input.exists()).toBeTruthy()
            cmp.vm.data.value = 'FALSE'
            input.vm.$emit('change')

            expect(cmp.emitted().change[0][0]).toEqual({
              type: 'Boolean',
              value: 'FALSE'
            })
          })
          it('emits change event if value is TRUE', () => {
            const input = cmp.findAll('.el-select').at(1)
            expect(input.exists()).toBeTruthy()
            cmp.vm.data.value = 'TRUE'
            input.vm.$emit('change')

            expect(cmp.emitted().change[0][0]).toEqual({
              type: 'Boolean',
              value: 'TRUE'
            })
          })
        })
      })

      describe('it can change the type to a number', () => {
        it('can change the type to boolean', () => {
          propsData.type = 'String'
          render()

          expect(cmp.vm.data.type).not.toEqual('Number')

          cmp.findAll('.el-select-dropdown__item').at(1).trigger('click')

          expect(cmp.vm.data.type).toEqual('Number')
        })

        describe('validates the field on change', () => {
          beforeEach(() => {
            propsData.type = 'Number'
            render()
          })
          it('shows no error if no input given', (done) => {
            cmp.destroy()
            render()

            const input = cmp.findAll('input').at(1)
            expect(input.exists()).toBeTruthy()
            input.setValue('')
            input.trigger('blur')
            input.trigger('change')

            cmp.vm.$nextTick(() => {
              expect(cmp.html()).toMatchSnapshot()
              const err = cmp.find('.el-form-item__error')
              expect(err.exists()).toBeFalsy()
              expect(cmp.emitted().change[0][0]).toEqual({
                type: 'Number',
                value: ''
              })
              done()
            })
          })
          it('shows error if invalid input', (done) => {
            render()
            const input = cmp.findAll('input').at(1)
            expect(input.exists()).toBeTruthy()
            input.setValue('string')
            input.trigger('blur')
            input.trigger('change')

            cmp.vm.$nextTick(() => {
              expect(cmp.html()).toMatchSnapshot()
              const err = cmp.find('.el-form-item__error')
              expect(err.exists()).toBeTruthy()
              expect(err.text()).toEqual('input_type.number')
              expect(cmp.emitted().change).toEqual(undefined)
              done()
            })
          })
          it('emits change event if value is a number', () => {
            const input = cmp.findAll('input').at(1)
            expect(input.exists()).toBeTruthy()
            input.setValue('12411')
            input.trigger('change')

            expect(cmp.emitted().change[0][0]).toEqual({
              type: 'Number',
              value: '12411'
            })
          })
          it('emits change event if value is a negative float number', () => {
            const input = cmp.findAll('input').at(1)
            expect(input.exists()).toBeTruthy()
            input.setValue('-3.141')
            input.trigger('change')

            expect(cmp.emitted().change[0][0]).toEqual({
              type: 'Number',
              value: '-3.141'
            })
          })
        })
      })
    })

    it('has a method to validate the form', () => {
      const fn = cmp.vm.validate
      expect(fn).toEqual(expect.any(Function))
      cmp.vm.$refs.inputTypeForm.validate = jest.fn()
      const cb = jest.fn()

      fn(cb)
      expect(cmp.vm.$refs.inputTypeForm.validate).toHaveBeenCalledWith(cb)
    })
  })
})
