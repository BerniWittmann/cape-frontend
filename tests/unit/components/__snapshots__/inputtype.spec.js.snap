// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components InputType has a prop to control the size 1`] = `
<form class="el-form">
  <div class="el-form-item el-form-item--small">
    <!---->
    <div class="el-form-item__content">
      <div class="input-with-select el-input el-input--small el-input-group el-input-group--prepend">
        <div class="el-input-group__prepend">
          <div class="el-select el-select--small">
            <!---->
            <div class="el-input el-input--small el-input--suffix">
              <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="请选择" class="el-input__inner">
              <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----></span>
              <!----></span>
              <!---->
            </div>
            <div class="el-select-dropdown el-popper" style="display: none;">
              <div class="el-scrollbar" style="display: none;">
                <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
                  <ul class="el-scrollbar__view el-select-dropdown__list">
                    <!---->
                    <li class="el-select-dropdown__item selected"><span>types.string</span></li>
                    <li class="el-select-dropdown__item"><span>types.number</span></li>
                    <li class="el-select-dropdown__item"><span>types.boolean</span></li>
                  </ul>
                </div>
                <div class="el-scrollbar__bar is-horizontal">
                  <div class="el-scrollbar__thumb" style="width: 0px; transform: translateX(0%); -webkit-transform: translateX(0%);"></div>
                </div>
                <div class="el-scrollbar__bar is-vertical">
                  <div class="el-scrollbar__thumb" style="height: 0px; transform: translateY(0%); -webkit-transform: translateY(0%);"></div>
                </div>
              </div>
              <p class="el-select-dropdown__empty">
                无数据
              </p>
            </div>
          </div>
        </div><input type="text" autocomplete="off" class="el-input__inner">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </div>
</form>
`;

exports[`Components InputType has a prop to hide the type select 1`] = `
<form class="el-form">
  <div class="el-form-item el-form-item--normal">
    <!---->
    <div class="el-form-item__content">
      <div class="input-with-select el-input el-input--normal">
        <!----><input type="text" autocomplete="off" class="el-input__inner">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </div>
</form>
`;

exports[`Components InputType it can change the type it can change the type to a boolean validates the field on change shows error if invalid input 1`] = `
<form class="el-form">
  <div class="el-form-item is-error el-form-item--normal">
    <!---->
    <div class="el-form-item__content">
      <div class="input-with-select el-input el-input--normal el-input-group el-input-group--prepend">
        <div class="el-input-group__prepend">
          <div class="el-select el-select--normal">
            <!---->
            <div class="el-input el-input--normal el-input--suffix">
              <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="请选择" class="el-input__inner">
              <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----></span><i class="el-input__icon el-input__validateIcon el-icon-circle-close"></i></span>
              <!---->
            </div>
            <div class="el-select-dropdown el-popper" style="display: none;">
              <div class="el-scrollbar" style="">
                <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
                  <ul class="el-scrollbar__view el-select-dropdown__list">
                    <!---->
                    <li class="el-select-dropdown__item"><span>types.string</span></li>
                    <li class="el-select-dropdown__item"><span>types.number</span></li>
                    <li class="el-select-dropdown__item selected"><span>types.boolean</span></li>
                  </ul>
                </div>
                <div class="el-scrollbar__bar is-horizontal">
                  <div class="el-scrollbar__thumb" style="transform: translateX(0%); -webkit-transform: translateX(0%);"></div>
                </div>
                <div class="el-scrollbar__bar is-vertical">
                  <div class="el-scrollbar__thumb" style="transform: translateY(0%); -webkit-transform: translateY(0%);"></div>
                </div>
              </div>
              <!---->
            </div>
          </div>
        </div><input type="text" autocomplete="off" class="el-input__inner">
        <!---->
        <!---->
        <!---->
      </div>
      <div class="el-form-item__error el-zoom-in-top-enter el-zoom-in-top-enter-active">
        input_type.boolean
      </div>
    </div>
  </div>
</form>
`;

exports[`Components InputType it can change the type it can change the type to a boolean validates the field on change shows error if no input given 1`] = `
<form class="el-form">
  <div class="el-form-item is-error el-form-item--normal">
    <!---->
    <div class="el-form-item__content">
      <div class="input-with-select el-input el-input--normal el-input-group el-input-group--prepend">
        <div class="el-input-group__prepend">
          <div class="el-select el-select--normal">
            <!---->
            <div class="el-input el-input--normal el-input--suffix">
              <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="请选择" class="el-input__inner">
              <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----></span><i class="el-input__icon el-input__validateIcon el-icon-circle-close"></i></span>
              <!---->
            </div>
            <div class="el-select-dropdown el-popper" style="display: none;">
              <div class="el-scrollbar" style="">
                <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
                  <ul class="el-scrollbar__view el-select-dropdown__list">
                    <!---->
                    <li class="el-select-dropdown__item"><span>types.string</span></li>
                    <li class="el-select-dropdown__item"><span>types.number</span></li>
                    <li class="el-select-dropdown__item selected"><span>types.boolean</span></li>
                  </ul>
                </div>
                <div class="el-scrollbar__bar is-horizontal">
                  <div class="el-scrollbar__thumb" style="transform: translateX(0%); -webkit-transform: translateX(0%);"></div>
                </div>
                <div class="el-scrollbar__bar is-vertical">
                  <div class="el-scrollbar__thumb" style="transform: translateY(0%); -webkit-transform: translateY(0%);"></div>
                </div>
              </div>
              <!---->
            </div>
          </div>
        </div><input type="text" autocomplete="off" class="el-input__inner">
        <!---->
        <!---->
        <!---->
      </div>
      <div class="el-form-item__error el-zoom-in-top-enter el-zoom-in-top-enter-active">
        input_type.required
      </div>
    </div>
  </div>
</form>
`;

exports[`Components InputType it can change the type it can change the type to a number validates the field on change shows error if invalid input 1`] = `
<form class="el-form">
  <div class="el-form-item is-error el-form-item--normal">
    <!---->
    <div class="el-form-item__content">
      <div class="input-with-select el-input el-input--normal el-input-group el-input-group--prepend">
        <div class="el-input-group__prepend">
          <div class="el-select el-select--normal">
            <!---->
            <div class="el-input el-input--normal el-input--suffix">
              <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="请选择" class="el-input__inner">
              <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----></span><i class="el-input__icon el-input__validateIcon el-icon-circle-close"></i></span>
              <!---->
            </div>
            <div class="el-select-dropdown el-popper" style="display: none;">
              <div class="el-scrollbar" style="">
                <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
                  <ul class="el-scrollbar__view el-select-dropdown__list">
                    <!---->
                    <li class="el-select-dropdown__item"><span>types.string</span></li>
                    <li class="el-select-dropdown__item selected"><span>types.number</span></li>
                    <li class="el-select-dropdown__item"><span>types.boolean</span></li>
                  </ul>
                </div>
                <div class="el-scrollbar__bar is-horizontal">
                  <div class="el-scrollbar__thumb" style="transform: translateX(0%); -webkit-transform: translateX(0%);"></div>
                </div>
                <div class="el-scrollbar__bar is-vertical">
                  <div class="el-scrollbar__thumb" style="transform: translateY(0%); -webkit-transform: translateY(0%);"></div>
                </div>
              </div>
              <!---->
            </div>
          </div>
        </div><input type="text" autocomplete="off" class="el-input__inner">
        <!---->
        <!---->
        <!---->
      </div>
      <div class="el-form-item__error el-zoom-in-top-enter el-zoom-in-top-enter-active">
        input_type.number
      </div>
    </div>
  </div>
</form>
`;

exports[`Components InputType it can change the type it can change the type to a number validates the field on change shows error if no input given 1`] = `
<form class="el-form">
  <div class="el-form-item is-error el-form-item--normal">
    <!---->
    <div class="el-form-item__content">
      <div class="input-with-select el-input el-input--normal el-input-group el-input-group--prepend">
        <div class="el-input-group__prepend">
          <div class="el-select el-select--normal">
            <!---->
            <div class="el-input el-input--normal el-input--suffix">
              <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="请选择" class="el-input__inner">
              <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----></span><i class="el-input__icon el-input__validateIcon el-icon-circle-close"></i></span>
              <!---->
            </div>
            <div class="el-select-dropdown el-popper" style="display: none;">
              <div class="el-scrollbar" style="">
                <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
                  <ul class="el-scrollbar__view el-select-dropdown__list">
                    <!---->
                    <li class="el-select-dropdown__item"><span>types.string</span></li>
                    <li class="el-select-dropdown__item selected"><span>types.number</span></li>
                    <li class="el-select-dropdown__item"><span>types.boolean</span></li>
                  </ul>
                </div>
                <div class="el-scrollbar__bar is-horizontal">
                  <div class="el-scrollbar__thumb" style="transform: translateX(0%); -webkit-transform: translateX(0%);"></div>
                </div>
                <div class="el-scrollbar__bar is-vertical">
                  <div class="el-scrollbar__thumb" style="transform: translateY(0%); -webkit-transform: translateY(0%);"></div>
                </div>
              </div>
              <!---->
            </div>
          </div>
        </div><input type="text" autocomplete="off" class="el-input__inner">
        <!---->
        <!---->
        <!---->
      </div>
      <div class="el-form-item__error el-zoom-in-top-enter el-zoom-in-top-enter-active">
        input_type.required
      </div>
    </div>
  </div>
</form>
`;

exports[`Components InputType it can change the type it can change the type to a string validates the field on change shows error if no input given 1`] = `
<form class="el-form">
  <div class="el-form-item is-error el-form-item--normal">
    <!---->
    <div class="el-form-item__content">
      <div class="input-with-select el-input el-input--normal el-input-group el-input-group--prepend">
        <div class="el-input-group__prepend">
          <div class="el-select el-select--normal">
            <!---->
            <div class="el-input el-input--normal el-input--suffix">
              <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="请选择" class="el-input__inner">
              <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----></span><i class="el-input__icon el-input__validateIcon el-icon-circle-close"></i></span>
              <!---->
            </div>
            <div class="el-select-dropdown el-popper" style="display: none;">
              <div class="el-scrollbar" style="">
                <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
                  <ul class="el-scrollbar__view el-select-dropdown__list">
                    <!---->
                    <li class="el-select-dropdown__item selected"><span>types.string</span></li>
                    <li class="el-select-dropdown__item"><span>types.number</span></li>
                    <li class="el-select-dropdown__item"><span>types.boolean</span></li>
                  </ul>
                </div>
                <div class="el-scrollbar__bar is-horizontal">
                  <div class="el-scrollbar__thumb" style="transform: translateX(0%); -webkit-transform: translateX(0%);"></div>
                </div>
                <div class="el-scrollbar__bar is-vertical">
                  <div class="el-scrollbar__thumb" style="transform: translateY(0%); -webkit-transform: translateY(0%);"></div>
                </div>
              </div>
              <!---->
            </div>
          </div>
        </div><input type="text" autocomplete="off" class="el-input__inner">
        <!---->
        <!---->
        <!---->
      </div>
      <div class="el-form-item__error el-zoom-in-top-enter el-zoom-in-top-enter-active">
        input_type.required
      </div>
    </div>
  </div>
</form>
`;

exports[`Components InputType renders 1`] = `
<form class="el-form">
  <div class="el-form-item el-form-item--normal">
    <!---->
    <div class="el-form-item__content">
      <div class="input-with-select el-input el-input--normal el-input-group el-input-group--prepend">
        <div class="el-input-group__prepend">
          <div class="el-select el-select--normal">
            <!---->
            <div class="el-input el-input--normal el-input--suffix">
              <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="请选择" class="el-input__inner">
              <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----></span>
              <!----></span>
              <!---->
            </div>
            <div class="el-select-dropdown el-popper" style="display: none;">
              <div class="el-scrollbar" style="">
                <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
                  <ul class="el-scrollbar__view el-select-dropdown__list">
                    <!---->
                    <li class="el-select-dropdown__item selected"><span>types.string</span></li>
                    <li class="el-select-dropdown__item"><span>types.number</span></li>
                    <li class="el-select-dropdown__item"><span>types.boolean</span></li>
                  </ul>
                </div>
                <div class="el-scrollbar__bar is-horizontal">
                  <div class="el-scrollbar__thumb" style="transform: translateX(0%); -webkit-transform: translateX(0%);"></div>
                </div>
                <div class="el-scrollbar__bar is-vertical">
                  <div class="el-scrollbar__thumb" style="transform: translateY(0%); -webkit-transform: translateY(0%);"></div>
                </div>
              </div>
              <!---->
            </div>
          </div>
        </div><input type="text" autocomplete="off" class="el-input__inner">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </div>
</form>
`;
